ambry-router/src/main/java/com/github/ambry/router/OperationController.java

```diff
@@ -106,13 +107,15 @@ public OperationController(String suffix, String defaultPartitionClass, AccountS
     this.responseHandler = responseHandler;
     this.nonBlockingRouter = nonBlockingRouter;
     // Warm up connections to dataNodes in local and remote DCs.
-    List<ResponseInfo> responseInfos = Collections.synchronizedList(new ArrayList<>());
     String localDatacenter = clusterMap.getDatacenterName(clusterMap.getLocalDatacenterId());
     Map<Boolean, List<DataNodeId>> localAndRemoteNodes = clusterMap.getDataNodeIds()
         .stream()
         // ignore any in-process "data nodes" without TCP ports
         .filter(dataNodeId -> dataNodeId.getPort() != DataNodeId.UNKNOWN_PORT)
         .collect(Collectors.partitioningBy(dataNodeId -> localDatacenter.equals(dataNodeId.getDatacenterName())));
+    // 'responseInfos' list needs to be thread safe since it it possible that the list can be modified as we are reading
+    // it for scenario where we returned from networkClient.warmUpConnections() due to time out.
+    List<ResponseInfo> responseInfos = new CopyOnWriteArrayList<>();
```

Comment:
What about other methods such as using a synchronized block, or waiting till all writer-threads have completed before iterating the list ?

Reply:
nvm, they may not work since the writes are scattered across code.